proposals/text/5629-file-similarity-retriever.md

```diff
@@ -0,0 +1,134 @@
+- Title: Adding FileSimilarityRetriever to haystack
+- Decision driver: @elundaeva
+- Start Date: 2023-08-28
+- Proposal PR: https://github.com/deepset-ai/haystack/pull/5629
+- Github Issue or Discussion: some links available in the "Detailed design" section below
+
+# Summary
+
+The retriever takes a file ID as query, searches for all documents from that file in the doc store and then performs a query for each document to find similar documents for each. Then these search results for each document are aggregated to produce a list of similar files.
+
+# Basic example
+
+The FileSimilarityRetriever would be instantiated as follows:
+
+``` python
+
+	  retriever = FileSimilarityRetriever(
+	      document_store = ElasticSearchDocumentStore,
+	      primary_retriever = EmbeddingRetriever, # defined separately, see full pipeline example below
+          top_k=30,
+	      file_aggregation_key = "file_id",
+          max_num_queries = 50
+	  )
+```
+
+And here is an example of how the node would work in the context of a full pipeline:
+
+``` yaml
+
+version: '1.19.0'
+name: 'FileSim'
+
+components:
+  - name: DocumentStore
+    type: DeepsetCloudDocumentStore
+    params:
+      similarity: cosine
+      embedding_dim: 768
+      return_embedding: false
+
+  - name: EmbeddingRetriever
+    type: EmbeddingRetriever
+    params:
+      document_store: DocumentStore
+      embedding_model: sentence-transformers/all-mpnet-base-v2
+      max_seq_len: 400
+      batch_size: 32
+      model_format: sentence_transformers
+      top_k: 50
+  - name: TextConverter # Converts files into documents
+    type: TextConverter
+  - name: FileSimilarityRetriever
+    type: FileSimilarityRetriever
+    params:
+      document_store: DocumentStore
+      primary_retriever: EmbeddingRetriever
```

Comment:
Note that I only used primary and secondary retriever because it wasn't possible to pass a list of retriever instances instead. Maybe it would be good to check again if this is now supported by pipelines. It would make the design more flexible

Reply:
I've discussed with Core and seems it hasn't changed yet & most likely won't in the near future. So, keeping primary & secondary retrievers as separate parameters seems like the way to go for now 👍

Reply:
That being said, @ZanSara mentioned that won't be an issue once Haystack 2.0 is out, so I can change it to list of retrievers at that point 🙂